go/WorkQueue.go

```diff
@@ -120,6 +121,72 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
 	return err
 }
 
+// AddItemAtomically to the work Queue.
+//
+// This creates a pipeline and executes it on the database if there is no other same item inside eitther main or processing queue.
+//
+// This method can be used only when database is locked
+func (workQueue *WorkQueue) AddItemAtomically(ctx context.Context, db *redis.Client, item Item) error {
+	txf := func(tx *redis.Tx) error {
+
+		pipeNew := db.Pipeline()
+
+		processingItemsInQueueCmd := pipeNew.LPos(ctx, workQueue.processingKey, item.ID, redis.LPosArgs{
+			Rank:   0,
+			MaxLen: 0,
+		})
+
+		workingItemsInQueueCmd := pipeNew.LPos(ctx, workQueue.mainQueueKey, item.ID, redis.LPosArgs{
+			Rank:   0,
+			MaxLen: 0,
+		})
+
+		_, err := pipeNew.Exec(ctx)
+
+		_, ProcessingQueueCheck := processingItemsInQueueCmd.Result()
+		_, WorkingQueueCheck := workingItemsInQueueCmd.Result()
+		if ProcessingQueueCheck == redis.Nil && WorkingQueueCheck == redis.Nil {
+
+			pipeEx := tx.Pipeline()
+			workQueue.AddItemToPipeline(ctx, pipeEx, item)
+			_, err = pipeEx.Exec(ctx)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	for {
+		err := db.Watch(ctx, txf, workQueue.mainQueueKey, workQueue.processingKey)
+		// Retry on transaction failure, return on anything else.
+		if err != redis.TxFailedErr {
+			return err
+		}
+	}
+}
+
+func (workQueue *WorkQueue) Lengths(ctx context.Context, db *redis.Client) (queueLen, processingLen int64, err error) {
+	tx := db.TxPipeline()
+
+	queueLen, err = tx.LLen(ctx, workQueue.mainQueueKey).Result()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	processingLen, err = tx.LLen(ctx, workQueue.processingKey).Result()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	_, err = tx.Exec(ctx)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return queueLen, processingLen, err
```

Comment:
No need to specify the values here. You can just use `return`. You specified them in the signature :)